### PR TITLE
feat: Rewrite buffer cache

### PR DESCRIPTION
**Pull request by driftluo**
Use a unified buffer abstract structure to manage the entire tentacle buffer to facilitate testing

dependence #248

### CODE REVIEW

File: src/buffer.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
+use crate::channel::mpsc::Sender as PrioritySender;
+use futures::channel::mpsc::Sender;
+use std::{
+    collections::VecDeque,
+    task::{Context, Poll},
+};
+
+/// If the buffer capacity is greater than u8 max, shrink it
```
Comment (doitian) [inline src/buffer.rs:8]: The comment is misleading.

```suggestion
/// If the buffer unused capacity is greater than u8 max, shrink it
```

File: src/buffer.rs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    pub fn try_send(&mut self, cx: &mut Context) -> SendResult {
+        while let Some(event) = self.high_buffer.pop_front() {
+            match self.sender.poll_ready(cx) {
+                Poll::Ready(Ok(())) => {
+                    if let Err(e) = self.sender.try_quick_send(event) {
+                        if e.is_full() {
+                            self.high_buffer.push_front(e.into_inner());
+                            return SendResult::Pending;
+                        } else {
+                            return SendResult::Disconnect;
```
Comment (doitian) [inline src/buffer.rs:66]: Why buffer is not cleared here?

File: src/buffer.rs [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+                        }
+                    }
+                }
+                Poll::Pending => {
+                    self.high_buffer.push_front(event);
+                    return SendResult::Pending;
+                }
+                Poll::Ready(Err(_)) => {
+                    self.high_buffer.clear();
```
Comment (doitian) [inline src/buffer.rs:75]: Why only clear the high buffer? Should normal buffer be cleared as well?

File: src/buffer.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+                            return SendResult::Disconnect;
+                        }
+                    }
+                }
+                Poll::Pending => {
+                    self.normal_buffer.push_front(event);
+                    return SendResult::Pending;
+                }
+                Poll::Ready(Err(_)) => {
+                    self.normal_buffer.clear();
```
Comment (doitian) [inline src/buffer.rs:97]: Why only clear the normal buffer? Should high buffer be cleared as well?

File: src/buffer.rs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+        }
+        while let Some(event) = self.normal_buffer.pop_front() {
+            match self.sender.poll_ready(cx) {
+                Poll::Ready(Ok(())) => {
+                    if let Err(e) = self.sender.try_send(event) {
+                        if e.is_full() {
+                            self.normal_buffer.push_front(e.into_inner());
+                            return SendResult::Pending;
+                        } else {
+                            return SendResult::Disconnect;
```
Comment (doitian) [inline src/buffer.rs:88]: Why buffer is not cleared here?

File: src/buffer.rs [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    pub fn try_send(&mut self, cx: &mut Context) -> SendResult {
+        while let Some(event) = self.buffer.pop_front() {
+            match self.sender.poll_ready(cx) {
+                Poll::Ready(Ok(())) => {
+                    if let Err(e) = self.sender.try_send(event) {
+                        if e.is_full() {
+                            self.buffer.push_front(e.into_inner());
+                            return SendResult::Pending;
+                        } else {
+                            return SendResult::Disconnect;
```
Comment (doitian) [inline src/buffer.rs:147]: Why buffer is not cleared here?

File: src/buffer.rs [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+        }
+        self.shrink_to_fit();
+        SendResult::Ok
+    }
+
+    pub fn take(&mut self) -> (Sender<T>, VecDeque<T>) {
+        (self.sender.clone(), ::std::mem::take(&mut self.buffer))
+    }
+
+    pub fn sender(&self) -> Sender<T> {
```
Comment (doitian) [inline src/buffer.rs:169]: It's better to explicitly show that this function clones the sender.

```suggestion
    pub fn clone_sender(&self) -> Sender<T> {
```

File: src/buffer.rs (hunk 2 above)
Comment (doitian) [inline src/buffer.rs:66]: Log the error?

File: src/buffer.rs [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+                        } else {
+                            return SendResult::Disconnect;
+                        }
+                    }
+                }
+                Poll::Pending => {
+                    self.high_buffer.push_front(event);
+                    return SendResult::Pending;
+                }
+                Poll::Ready(Err(_)) => {
```
Comment (doitian) [inline src/buffer.rs:74]: Log the error?

File: src/buffer.rs (hunk 5 above)
Comment (doitian) [inline src/buffer.rs:88]: Log the error?

File: src/buffer.rs [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+                        } else {
+                            return SendResult::Disconnect;
+                        }
+                    }
+                }
+                Poll::Pending => {
+                    self.normal_buffer.push_front(event);
+                    return SendResult::Pending;
+                }
+                Poll::Ready(Err(_)) => {
```
Comment (doitian) [inline src/buffer.rs:96]: Log the error?

File: src/buffer.rs (hunk 6 above)
Comment (doitian) [inline src/buffer.rs:147]: Log the error?

File: src/buffer.rs [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+                        } else {
+                            return SendResult::Disconnect;
+                        }
+                    }
+                }
+                Poll::Pending => {
+                    self.buffer.push_front(event);
+                    return SendResult::Pending;
+                }
+                Poll::Ready(Err(_)) => {
```
Comment (doitian) [inline src/buffer.rs:155]: Log the error?

File: src/buffer.rs (hunk 2 above)
Comment (driftluo) [reply src/buffer.rs:66]: There is no log inside the structure definition, only when it is used, it is necessary to consider whether to add

Comment (doitian) [reply src/buffer.rs:66]: If it's OK to add a String in `SendResult::Disconnect`, we can add the reason in it.

Comment (driftluo) [reply src/buffer.rs:66]: There is only one reason for disconnect, the sender end is closed/dropped

File: src/substream.rs [hunk 11]
```diff
@@ -215,16 +208,11 @@ where
         }
 
         if self.session_proto_sender.is_some() {
-            self.session_proto_buf
-                .push_back(SessionProtocolEvent::Closed);
-
+            let (mut sender, mut events) = self.session_proto_sender.take().unwrap().take();
```
Comment (TheWaWaR) [inline src/substream.rs:211]:
```rust
if let Some(session_proto_sender) = self.session_proto_sender.take() {
     // ...
}
```

**Pull request merged**